### ISSUE TITLE
Add Linux 64bit VST for next release

### ISSUE THREAD

**Issue by dab3-2014** [asb2m10/dexed#50]
Could you please provide any documentation on how to install the VST on Linux? 

I tried to compile it with a simple `$ make` but it errors as it can not locate the VST SDK folder. Is there any way to declare the VST SDK path in the `$ make` command? Is editing the MAKEFILE an option, although there's a warning against this in the preamble.

**Comment by asb2m10** [comment]
Sorry for the late reply. I don't see the problem of changing the 'hardcorded' VST SDK path. 

I'm planning to add a Wiki page specifically for Linux. I'll keep that issue open until the dexed wiki page is done.

**Comment by SpotlightKid** [comment]
The way I do it in the [dexed-vst-git](https://aur.archlinux.org/packages/dexed-vst-git/) AUR package:

```
srcdir=/path/to/where/i/extracted/vst3sdk
cd Builds/Linux
sed -i -e "s|-I ~/src/vstsdk2.4|-I $srcdir/VST3\\\\ SDK|" Makefile
make CONFIG=Release
install -Dm755 build/Dexed.so /usr/lib/vst/Dexed.so
```

**Comment by asb2m10** [comment]
Thanks for the recipe @SpotlightKid ... I've added it to the Linux Wiki build: https://github.com/asb2m10/dexed/wiki/Linux-Build

**Comment by Learn-Digital-Audio** [comment]
Hello. That link goes to a blank page. Not all Linux users are able to build from source, or want to. I don't quite know why there is compiled versions for Windows and Mac, but not Linux, it's very frustrating, especially on a FOSS project of such excellent quality!

Would it be possible for there to be an official 64bit Linux build on the download page? Thanks!

**Comment by SpotlightKid** [comment]
The problem with Linux binaries is, that there is no *one* Linux, they all differ in the versions of libraries they have, the C compiler they use and the desktop integration they offer (the last is not usually an issue for plug-ins, though).

So a binary compiled for e.g. Debian/Ubuntu is not guaranteed to run e.g. on Arch Linux, or may behave in slightly different ways, hard to debug ways. Also, people usually want to install software on Linux via the system's packaging system, so installed software can be tracked, updated automatically and easily removed.

Nowadays, though, it should be easier to provide pre-compiled binaries for different Linux flavours, using build services based on Docker or similar. But setting this up and maintaining it still needs a lot of work and public build services (e.g. Travis) often are poorly equipped for building and testing audio software.

**Comment by Learn-Digital-Audio** [comment]
Thanks for your detailed reply. Regarding there not being one Linux and different distros having different dependencies, it seems many Linux VST plugins are just distributed as one .so file which works everywhere? For example, U-he have just one .so file for all different distributions.

It would be great if the Dexed home page could at least have detailed info on how to compile the source for the most popular distros, if not a compiled .so file. I know its work for someone and everyone is doing this for free, but there is no harm putting in a request :)

Afraid Travis looks beyond my abilities, but maybe it's not so hard for someone else to use it and make a binary?

Maybe approaching different repos and asking them to include Dexed would be a real positive too. For example http://rpmfusion.org/ if not the official Fedora one.

I found a version compiled from 2014 here, but that's ancient. I wrote to the guy to see if he would update it but have not heard back:

http://distrho.sourceforge.net/ports

I found this Fedora related site that might help someone build/host Dexed:

https://copr.fedorainfracloud.org/coprs/

**Comment by maxxatgit** [comment]
Greetings from yet another Linux user;
I'm running Ubuntu 16.04 64bit . I tried the recipe, and it didn't work. I keep getting this error when I run: make CONFIG=Release
Compiling PluginFx.cpp
/bin/sh: 1: Syntax error: Unterminated quoted string
Makefile:111: recipe for target 'build/intermediate/Release/PluginFx_d84f776e.o' failed
make: *** [build/intermediate/Release/PluginFx_d84f776e.o] Error 2

I downloaded and unzipped the vst3 sdk and tried doing the recipe exactly as shown (replacing the "/path/to/where/i/extracted/vst3sdk" with the location of the unzipped vst folder), and I tried replacing $srcdir with  the location of the unzipped vst folder, and that produced the same error.

I'm no linux sysadmin, but I'm not a noob either. I can often find solutions to compiler issues by googling the errors. I  found what an "Unterminated quoted string" is, and I searched the Makefile for a missing closing quote, but they are all matched pairs. The build directories are present, but empty so there is nothing to search in there...

I have to repeat Learn-Digital-Audio's comments; Linux Audio plugins are just compiled .so files. So, SpotlightKid, can you compile a .so file for the latest Dexed version and post it here? Then maybe asb2m10 can add a Linux downloadable file to go with the Windows and MacOS downloads.

The 2014 version sounds great! It would be really awesome if you folks could make the current version of this fine software more readily available to Linux users.

Thanks in advance!

**Comment by SpotlightKid** [comment]
You probably have a space or another special character in the path `srcdir`, which messes up the sed command or the Makefile. Don't do that.

**Comment by Learn-Digital-Audio** [comment]
I extracted the .so file right out of the kxstudio repos:

https://launchpad.net/~kxstudio-debian/+archive/ubuntu/plugins/+packages

see "distrho-ports - 2:20160323-1kxstudio2", just download the one you want,  and extract it directly, for example:

dexed-lv2_20160323-1kxstudio2_amd64.deb

but even this version is not the latest, but at least it's this year 2016-03-23

i spent hours trying to get this simple info on how to get a recent version of this software! even this support thread was unable to supply it! i had to write to a trusted linux audio coder directly and ask his opinion before i found the answer... why would anyone bother to jump this many hoops?!

i have been exploring the world of linux audio all year, and am finding many frustrations that are fundamentally caused by a massive lack of leadership and organisation. everyone is doing their own thing, nobody wants to join under one organisation. its the same all over linux.

linux's strength is currently also it's weakness. we need someone like red hat to throw resources at stuff to get things done. technically linux can dominate audio tomorrow, but does anyone even want that to happen?

**Comment by asb2m10** [comment]
The site is updated to https://github.com/asb2m10/dexed/wiki/Linux-Builds

Please don't tell me it is massive lack of leadership: let me remind you that I have a full time job; managers that asks me to work on weekends, a family with a young kid, so when I DO have free time, I decide what I want to do. There is a reason I don't accept donation and it is because I don't want somebody to give me money to then tell me what to do NOW on my project. It will take more time but it will be done. 

I already do 4 builds per release (MAC VST, MAC AU, WIN VST32, WIN VST64); I will keep it that way. I don't mind releasing the 64-bit version for Linux couple weeks later, but I can't guarantee that it will work on all distributions and don't want to support multiple builds/distributions. If somebody knows the exact linker arguments to build a very fat binary like u-he does for Linux; I would be more than happy to know them and use them in my next release.

If there is a distribution maintainer that ask me do add files to generate their package artifact, I will be more than happy to cooperate.

**Comment by Learn-Digital-Audio** [comment]
asb2m10, I was talking in general about Linux software, not you personally. Thank you for all the time you put in for free, I do appreciate it!

Sorry if i came across in a negative way. It's just very frustrating coming over from Windows trying to get a working audio production system. I had no idea how great Linux was until I used it for a while, now Windows just seems horrible in comparison. It would be a dream to get all the tools from Windows working well in Linux, but it's out of my hands until I learn to code. All I can do is go around and try to promote Linux and hope people try it and love it like me.

**Comment by trebmuh** [comment]
> It would be a dream to get all the tools from Windows working well in Linux

No, it would be hell. You certainly are thinking that because you just switched and you're still on the learning curve. Advice is to wait one year or two before making such statements. You'll then have experimented this OS and your comments will be much more relevant.

**Comment by Learn-Digital-Audio** [comment]
@trebmuh, I think i can decide for myself if it would be good or bad to have all the tools i have been using successfully for years working well on Linux. Most of them are already, Libre Office, Firefox, Blender, Filezilla, VLC Player, VScode... even Photoshop and Foobar are 100% fine in Wine... Reaper is on the way over to Linux too. I can't understand your comment at all.

**Comment by trebmuh** [comment]
You didn't precise you're talking for yourself. Then, as a common thing, I answered that it would be hell to get the tools from windows into linux. I maintain that.

For Libre Office, Firefox, ... it's not about porting windows tools to linux as well, that's actually the opposite.

**Comment by asb2m10** [comment]
I've added the next Linux build in the beta, please check if this works on your side https://github.com/asb2m10/dexed/releases/tag/v0.9.3b2

Usually do an '$ ldd Dexed.so' to confirm that you have all the required libraries before running it.

Thanks.

**Comment by Learn-Digital-Audio** [comment]
Thanks for the Linux version!

I am on Fedora 24. Here is my result from ldd, which seems all good:

>  ldd Dexed.so 
> 	linux-vdso.so.1 (0x00007ffffa75c000)
> 	libX11.so.6 => /lib64/libX11.so.6 (0x00007fc4413a2000)
> 	libXext.so.6 => /lib64/libXext.so.6 (0x00007fc441190000)
> 	libXinerama.so.1 => /lib64/libXinerama.so.1 (0x00007fc440f8c000)
> 	libasound.so.2 => /lib64/libasound.so.2 (0x00007fc440c95000)
> 	libdl.so.2 => /lib64/libdl.so.2 (0x00007fc440a91000)
> 	libfreetype.so.6 => /usr/lib64/freetype-freeworld/libfreetype.so.6 (0x00007fc4407e1000)
> 	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fc4405c5000)
> 	librt.so.1 => /lib64/librt.so.1 (0x00007fc4403bd000)
> 	libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007fc440034000)
> 	libm.so.6 => /lib64/libm.so.6 (0x00007fc43fd2b000)
> 	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fc43fb14000)
> 	libc.so.6 => /lib64/libc.so.6 (0x00007fc43f751000)
> 	libxcb.so.1 => /lib64/libxcb.so.1 (0x00007fc43f52f000)
> 	/lib64/ld-linux-x86-64.so.2 (0x000055fb3ce0c000)
> 	libz.so.1 => /lib64/libz.so.1 (0x00007fc43f319000)
> 	libbz2.so.1 => /lib64/libbz2.so.1 (0x00007fc43f108000)
> 	libpng16.so.16 => /lib64/libpng16.so.16 (0x00007fc43eed5000)
> 	libXau.so.6 => /lib64/libXau.so.6 (0x00007fc43ecd1000)

It loads into the new Reaper 5.351 Linux preview version fine, and sound is coming out fine :)

I have found that when I am in the CART browser I can't load the cart by pressing the LOAD button. I have to go to a new cart and press enter before it loads. Here is a screen capture showing what I mean:

http://www.learndigitalaudio.com/wp-content/uploads/2017/03/Screencast-from-12-03-17-110421.webm

The loading is happening when I press enter.

The only other issue I have encountered so far is maybe because I am using a Wacom pen rather than a mouse. I can't actually select a program to load from a cartridge by clicking it:

http://www.learndigitalaudio.com/wp-content/uploads/2017/03/wacom.jpg

I have to hold down on the program name, wiggle the selection until it selects (with a green inverted background) and goes into drag and drop mode. Now I can move the program around, and with a lot of fiddling about get it to load, but often it won't load. Is there a keyboard shortcut maybe?

I expect my graphics tablet use is a minority one, so don't imagine you have the time to fix it... but if there is any obvious fix then great!

Thanks again, all the best!

**Comment by asb2m10** [comment]
Happy to see that it works.

When you click on load, it should open a file dialog and with JUCE there was an issue that seems to be fixed with the 4.3.1 version. The next version of Dexed will probably use this JUCE version. Otherwise on the cartridge sysex file, you need to double click it in order to load them.
